[Bug] GoBig list index out of range / various other fixes for issues

devilismyfriend opened the pull request:
Issue described here:
https://github.com/hlky/stable-diffusion-webui/issues/464

not sure exactly who commented out the flag on the process_image() but this will fix it and added a small comment

devilismyfriend reviewed (commented):
This is it

codedealer commented:
> Issue described here: #464
> 
> not sure exactly who commented out the flag on the process_image() but this will fix it and added a small comment

it was here 1922a44334bc6ff22e930fa5601be72f6c75e694 to fix image duplication in output. Did you check that it's not bringing back this behavior?

lozenge commented:
The default value for `imgProcessorTask` in `process_image()` is currently set to `True`. Setting this to `False` avoids the duplicate image output.

devilismyfriend commented:
> The default value for `imgProcessorTask` in `process_image()` is currently set to `True`. Setting this to `False` avoids the duplicate image output.

Thanks just noticed it myself, not sure why it was set to True, or when was it changed but it should be good now

devilismyfriend commented:
<img width="616" alt="image" src="https://user-images.githubusercontent.com/87043616/188065317-1047e11a-c99d-4c21-9ed8-fcbc9efd1e98.png">
added increment buttons to sampler so it's easier to fine tune when using sensitive samplers like k_euler_a

Cainor commented:
This thread fixes the GoBig out of index error? If so, I want to thank you for your work.

devilismyfriend commented:
> This thread fixes the GoBig out of index error? If so, I want to thank you for your work.

No Problem.
This PR actually has quite a lot of changes, but the GoBig fix is one of them.

JoshuaKimsey commented:
> added increment buttons to sampler so it's easier to fine tune

Thank you @devilismyfriend for this! On a 4K display, it's been a bit of a challenge to get it dialed in right for the correct step size, especially when doing the benchmarking. Any chance you could implement this for the other sliders too? Such as CFG scale, now that it goes into the negative?

altryne commented:
![CleanShot 2022-09-02 at 16 28 45@2x](https://user-images.githubusercontent.com/463317/188242610-8b8b71a4-0ffe-4b2f-b67d-d8a29c783bad.jpg)
I don't love how this looks, it's inconsistent with the other design language IMO

I agree that it's hard to finetune 1-2 steps, to this end dynamic variables were introduced.

altryne commented on README.md:
```diff
@@ -2,6 +2,9 @@
 # This repo is for development, there may be bugs and new features
 
+# Notice
+- To use the new LDSR settings please make sure to re-clone the LDSR repo.
```
Please expand a little bit on this?

devilismyfriend replied:
The idea is to clone the LDSR repo again to get the new LDSR module that takes the arguments (instruction are a bit further down with the other upscalers), how would yo suggest rewriting it?

altryne commented on frontend/frontend.py:
```diff
@@ -62,9 +63,14 @@ def draw_gradio_ui(opt, img2img=lambda x: x, txt2img=lambda x: x,imgproc=lambda
                                     _js='(x) => navigator.clipboard.writeText(x)', fn=None, show_progress=False)
                             output_txt2img_stats = gr.HTML(label='Stats')
                     with gr.Column():
+                        with gr.Row():
```
Big no on this, this complicates the UI and can be done with dynamic params (which needs documentation)

devilismyfriend replied:
I agree that it isn't the best looking solution (although with the added styles.css it looked much worse lol), and while I agree dynamic params are powerful, I don't think they're really a replacment for just a +- button, reason is, if you're doing dynamic params you're treating the UI more like CLI, which in that case you're probably trying to minimize button presses etc, to me tho, even if I can do CLI I'd rather use the buttons (I like to just use a mouse while I generate hundreds of photos lol) so I'd rather use buttons.

Not saying which way is correct or whether we should or shouldn't remove it, I just think this should be an option, other users that saw this already asked for it to be also added to the CFG parameter cause like the steps, it's hard to control.

Maybe we change this into an option and set the non-button slider as the default?

altryne replied:
dynamic params isn't the best solution for sure, but gradio is restrictive, we use sliders because for most things they are most useful. I wish they had an option to arbitrarily adding params, but they don't 
Adding buttons like this (outside of the bounding box of the slider) is just not looking great and is not a great solution, as I already saw the same users ask for this everywhere.

This tab has a nice looking ui, and it's consistent, I would like to keep it that way

devilismyfriend replied:
> dynamic params isn't the best solution for sure, but gradio is restrictive, we use sliders because for most things they are most useful. I wish they had an option to arbitrarily adding params, but they don't Adding buttons like this (outside of the bounding box of the slider) is just not looking great and is not a great solution, as I already saw the same users ask for this everywhere.
> 
> This tab has a nice looking ui, and it's consistent, I would like to keep it that way

It's the same old question of design vs function I guess, you sure we shouldn't just add it as an option and not set it to default?

JoshuaKimsey replied:
I do agree having the controls lay outside the bounding box is an inconsistent look. Couldn't the "-" and "+" be placed inside of the bounding box, possibly at the beginning and end of the slider bar itself? That would make it more consistent with the UI and also make it more obvious what they do. 

Or is that just not something Gradio will allow due to how it functions?

devilismyfriend replied:
unfortunately that isn't an option, if it was I'd implemented like that haha

altryne commented:
> > This thread fixes the GoBig out of index error? If so, I want to thank you for your work.
> 
> No Problem. This PR actually has quite a lot of changes, but the GoBig fix is one of them.

Thanks for the amazing work @devilismyfriend ,I've added some comments. 

Please try to scope changes in PRs, I missed these as the title said "[BUG] GoBig list index" but this actually had other UI tweaks that are unrelated, which need a separate discussion.

altryne commented on README.md:
```diff
@@ -62,6 +65,15 @@ into the `/stable-diffusion/src/gfpgan/experiments/pretrained_models` directory.
+  - Git clone https://github.com/devilismyfriend/latent-diffusion into your stable-diffusion-main/src/ folder
+  - Rename latent-diffusion-main folder to latent-diffusion
+  - If on windows: run download_models.bat to download the required model files
+  - Otherwise to manually install the model download [project.yaml](https://heibox.uni-heidelberg.de/f/31a76b13ea27482981b4/?dl=1) and [last.cpkt](https://heibox.uni-heidelberg.de/f/578df07c8fc04ffbadf3/?dl=1) and rename last.ckpt to model.ckpt
```
Where are these links hosted? Could they fail under load? 
Asking cause, I want to add an auto downloader of moderls to the colab version of this

HenkDz replied:
why no changes are done for the past 24 hours? is @hlky okey?

altryne replied:
The man needs sleep lol

devilismyfriend replied:
Looks like the university of Heidelberg lol, idk, it's the links I got off the latent diffusion repo which are probably provided by compVis

devilismyfriend commented:
> > added increment buttons to sampler so it's easier to fine tune
> 
> Thank you @devilismyfriend for this! On a 4K display, it's been a bit of a challenge to get it dialed in right for the correct step size, especially when doing the benchmarking. Any chance you could implement this for the other sliders too? Such as CFG scale, now that it goes into the negative?

lol I'm also on a 4k monitor (OLED LG FTW), maybe it's just us haha

devilismyfriend commented:
> > > This thread fixes the GoBig out of index error? If so, I want to thank you for your work.
> > 
> > 
> > No Problem. This PR actually has quite a lot of changes, but the GoBig fix is one of them.
> 
> Thanks for the amazing work @devilismyfriend ,I've added some comments.
> 
> Please try to scope changes in PRs, I missed these as the title said "[BUG] GoBig list index" but this actually had other UI tweaks that are unrelated, which need a separate discussion.

Yeah I know it's all confusing, sorry, it's my first time collabing on github so there's a bit of learning curve..

altryne commented:
I finally was able to download these in colab, and if we're fixing the tab, these are very confusing: 
Fix Faces setting applies ONLY to fix faces but is taking over a whole tab, where most of the "mode" settings, is in another hidden tab making the user have to click

Then in the other tab, the options are all in disarray and very not visually consistent 
I would recommend taking another stab, like so: 

arrange the specific settings for each mode (face, ESR, LSDR) in their own gr.Group(visible=True)
Then take out the settings out of the tab, and make it so the specific options show up when an option is checked with binding a .change function. 
 

![CleanShot 2022-09-02 at 17 26 16@2x](https://user-images.githubusercontent.com/463317/188245939-933c85a8-9d4a-41a7-a5c6-ab926da376d2.jpg)

altryne commented:
let me know if this is unrealted, or if was fixed here, but just got this inside colab when trying to upscale
![CleanShot 2022-09-02 at 17 37 23@2x](https://user-images.githubusercontent.com/463317/188246618-b2df2965-76ea-444c-937c-5b2b57374e2f.jpg)

devilismyfriend commented:
> I finally was able to download these in colab, and if we're fixing the tab, these are very confusing: Fix Faces setting applies ONLY to fix faces but is taking over a whole tab, where most of the "mode" settings, is in another hidden tab making the user have to click
> 
> Then in the other tab, the options are all in disarray and very not visually consistent I would recommend taking another stab, like so:
> 
> arrange the specific settings for each mode (face, ESR, LSDR) in their own gr.Group(visible=True) Then take out the settings out of the tab, and make it so the specific options show up when an option is checked with binding a .change function.
> 
> ![CleanShot 2022-09-02 at 17 26 16@2x](https://user-images.githubusercontent.com/463317/188245939-933c85a8-9d4a-41a7-a5c6-ab926da376d2.jpg)

Yeah this was the building blocks, I actually wanted each upscaler to have a tab for the settings to avoid confusion and clutter but gradio didn't play ball.

in regards to the +- minus buttons, I think maybe a custom css loader could also be useful, I'd personally want a design more like this

<img width="1473" alt="image" src="https://user-images.githubusercontent.com/87043616/188246670-831fcdc2-2b0f-4b3c-bbaf-99e0d3ad4f12.png">

devilismyfriend commented:
> > I finally was able to download these in colab, and if we're fixing the tab, these are very confusing: Fix Faces setting applies ONLY to fix faces but is taking over a whole tab, where most of the "mode" settings, is in another hidden tab making the user have to click
> > Then in the other tab, the options are all in disarray and very not visually consistent I would recommend taking another stab, like so:
> > arrange the specific settings for each mode (face, ESR, LSDR) in their own gr.Group(visible=True) Then take out the settings out of the tab, and make it so the specific options show up when an option is checked with binding a .change function.
> > ![CleanShot 2022-09-02 at 17 26 16@2x](https://user-images.githubusercontent.com/463317/188245939-933c85a8-9d4a-41a7-a5c6-ab926da376d2.jpg)
> 
> Yeah this was the building blocks, I actually wanted each upscaler to have a tab for the settings to avoid confusion and clutter but gradio didn't play ball.
> 
> in regards to the +- minus buttons, I think maybe a custom css loader could also be useful, I'd personally want a design more like this
> 
> <img alt="image" width="1473" src="https://user-images.githubusercontent.com/87043616/188246670-831fcdc2-2b0f-4b3c-bbaf-99e0d3ad4f12.png">

This is unrelated to here but I need to fix that on the LDSR repo, will push it now ;) thanks!

altryne commented:
>'d personally want a design more like this 
This is getting out of scope, for sure of the bug fix PR. Please revert this specific change 🙏

altryne commented:
I just found another issue, every time I upscale, every model, including the original SD model, are reloaded into memory again 
(not on this branch fix, but on main)
Can you address this? 
![CleanShot 2022-09-02 at 17 43 03@2x](https://user-images.githubusercontent.com/463317/188246840-a7699474-2ad7-41ad-b39f-eabc93485f4e.jpg)

devilismyfriend commented:
> I just found another issue, every time I upscale, every model, including the original SD model, are reloaded into memory again (not on this branch fix, but on main) Can you address this? ![CleanShot 2022-09-02 at 17 43 03@2x](https://user-images.githubusercontent.com/463317/188246840-a7699474-2ad7-41ad-b39f-eabc93485f4e.jpg)

That's by design, would you like it not to unload SD when usuing RealESRGAN as upscaler? it was added to make room on the VRAM for larger pictures/weaker hardware

devilismyfriend commented:
> 

pushed a fix for this to the LDSR repo

devilismyfriend commented:
> > 'd personally want a design more like this
> 
> This is getting out of scope, for sure of the bug fix PR. Please revert this specific change 🙏

Cool, I'm gonna try that group idea, seems like a cool dynamic way to display the settings

altryne commented:
> > I just found another issue, every time I upscale, every model, including the original SD model, are reloaded into memory again (not on this branch fix, but on main) Can you address this? ![CleanShot 2022-09-02 at 17 43 03@2x](https://user-images.githubusercontent.com/463317/188246840-a7699474-2ad7-41ad-b39f-eabc93485f4e.jpg)
> 
> That's by design, would you like it not to unload SD when usuing RealESRGAN as upscaler? it was added to make room on the VRAM for larger pictures/weaker hardware

Ah wow... I mean it makes sense now that you say this, but maybe don't reload the SD each time automatically? Only the next time we start a SD process? 

Not sure what's best here as I didn't do hardware tests (in colab) but models loading and unloading could take a while, but this happened when I tried to upscale

devilismyfriend commented:
> > > I just found another issue, every time I upscale, every model, including the original SD model, are reloaded into memory again (not on this branch fix, but on main) Can you address this? ![CleanShot 2022-09-02 at 17 43 03@2x](https://user-images.githubusercontent.com/463317/188246840-a7699474-2ad7-41ad-b39f-eabc93485f4e.jpg)
> > 
> > 
> > That's by design, would you like it not to unload SD when usuing RealESRGAN as upscaler? it was added to make room on the VRAM for larger pictures/weaker hardware
> 
> Ah wow... I mean it makes sense now that you say this, but maybe don't reload the SD each time automatically? Only the next time we start a SD process?
> 
> Not sure what's best here as I didn't do hardware tests (in colab) but models loading and unloading could take a while, but this happened when I tried to upscale.
> 
> Since you returned the checkboxes to run GFPGan and ESRGan to the main tab, shouldn't they be loaded together with the SD model? Otherwise how would they work?

SD, GFPGAN and RealESRGAN are always loaded on startup (should be changed imo), which is why those checkboxes will still work, that being said it can be changed to just reload them when using the checkboxes.
I can certainly use the loader/unloader module to make sure SD is loaded before generation etc.

in the topic of the groups in gradio, fun times, visible doesn't work lol docs say it should, but it doesn't

altryne commented:
>SD, GFPGAN and RealESRGAN are always loaded on startup (should be changed imo), which is why those checkboxes will still work, that being said it can be changed to just reload them when using the checkboxes.
I can certainly use the loader/unloader module to make sure SD is loaded before generation etc. 

So what unloads and reloads?

devilismyfriend commented:
> > SD, GFPGAN and RealESRGAN are always loaded on startup (should be changed imo), which is why those checkboxes will still work, that being said it can be changed to just reload them when using the checkboxes.
> > I can certainly use the loader/unloader module to make sure SD is loaded before generation etc.
> 
> So what unloads and reloads?

well before I added the function to do it (LDSR requires SD to be out of memory or else it won't work) they were just kept there all the time, now it's handled by ModelLoader, one of the reason I added it was to avoid global declaration at every turn.

devilismyfriend commented:
So changes in according to the review:

-commented out the added buttons for now for further discussion
-Image lab settings are now visible in accordance to the selection
-better models handling, only SD is loaded at boot with all the other loaded on demand but kept in memory in the case of a batch etc (unless it's a tiered upscaling like goLatent where SD is being unloaded at various points)

hlky commented:
@devilismyfriend i think altryne is away for a week from today, so let me know when you're done with this, thanks

devilismyfriend commented:
> @devilismyfriend i think altryne is away for a week from today, so let me know when you're done with this, thanks

it's done

devilismyfriend commented:
seems like there's some conflicts, I'll close them now
